datasync/chaindatafetcher/kafka/consumer.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2020 The klaytn Authors
+// This file is part of the klaytn library.
+//
+// The klaytn library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package kafka
+
+import (
+	"context"
+
+	"github.com/Shopify/sarama"
+)
+
+type Consumer struct {
+	cancel   chan bool
+	handler  map[string]func(*sarama.ConsumerMessage) error
+	consumer sarama.ConsumerGroup
+	ctx      context.Context
+	isActive bool
+}
+
+func NewConsumer(ctx context.Context, consumer sarama.ConsumerGroup) *Consumer {
+	return &Consumer{
+		cancel:   make(chan bool),
+		handler:  map[string]func(*sarama.ConsumerMessage) error{},
+		ctx:      ctx,
+		consumer: consumer,
+		isActive: false,
+	}
+}
+
+func (r *Consumer) Subscribe(topic string, handler func(*sarama.ConsumerMessage) error) error {
+	if r.handler[topic] != nil {
+		return nil
+	}
+	r.handler[topic] = handler
+
+	if r.isActive {
+		r.cancel <- true
```

Comment:
What does this line for? Doesn't it always close `for` loop in the below if there was no other `for` loops?

Reply:
Initially, the variable is set to `false`, so it won't close the `for` loop. The `Consume` method calls `Setup`, `ConsumeClaim`, and `Cleanup` consecutively. 
- `Setup` sets `isActive` be true.
- `ConsumeClaim` method has a `for` loop for fetching message from a channel. 
- `Cleanup` sets `isActive` be false.

So, when `Subscribe` is called while processing messages, it terminates the loop and create another loop to process messages.